tests/testthat/test-examples.R

```diff
@@ -0,0 +1,38 @@
+# this test requires a `man` directory in the `tests/testthat` directory
+# (presumably symlinked to the package root `man` directory to avoid duplication)
+# this also requires `devtools::document()` to be run before running the tests
+
+rd_files <- function() {
+  list.files(testthat::test_path("man"), pattern = "\\.[Rr]d$", full.names = TRUE)
+}
+
+suppress_warnings <- function(expr, pattern = "*", ...) {
+  withCallingHandlers(expr, warning = function(w) {
+    if (grepl(pattern, conditionMessage(w))) {
+      invokeRestart("muffleWarning")
+    }
+  })
+}
+
+for (i in rd_files()) {
+  testthat::with_mocked_bindings(
+    testthat::test_that(
+      paste0("example-", basename(i)),
+      {
+        testthat::skip_on_cran()
+        testthat::expect_no_error(
+          # surpress warnings coming from saving qenv https://github.com/insightsengineering/teal.code/issues/194
+          suppress_warnings(
+            pkgload::run_example(i, run_donttest = TRUE, run_dontrun = FALSE, quiet = TRUE),
+            "may not be available when loading"
+          )
+        )
+      }
+    ),
+    runApp = function(appDir, ...) { # nolint object_name_linter.
```

Comment:
```suggestion
    runApp = function(appDir, ...) { # nolint object_name_linter.
```

Shouldn't we mocked `shinyApp` instead of `runApp` since we replaced all the `runApp` with `shinyApp` in the examples?

Reply:
It really comes down to what function does what.
`shinyApp` returns an application object. This can theoretically could be modified afterwards before the application initiates, e.g. `x <- shinyApp(...); x_new <- foo(x); x_new`
`runApp` initiates the app in the sense of opening port, workers, servers etc. Note that `runApp()` is called inside `print.shiny.appobj` and `shiny.appobj` instance is returned by `shinyApp()`. This is why all the applications _works_ - when we put `x <- shunyApp(...); x` (or `shinyApp(...)` directly) we do (i) create an app object and also (ii) call print method on it.

On that note I think that we should mock `runApp` to step into initialization of the ready-to-be-initiated app instead of app object creation.

Reply:
Please correct me if I'm wrong.

The purpose of `with_mocked_bindings` is to temporarily redefine a function so that it behaves differently. You redefined how `runApp` should run, but we never use `runApp` in the examples section for each module. This suggests that the redefined function is never actually executed.

For example, I added `print` statement to your code
```r
    runApp = function(appDir, ...) { # nolint object_name_linter.
      print("runApp is called") # add print statement
      app <- shinytest2::AppDriver$new(appDir)
      app$stop()
    },
```

But I don't see this print statement when running `devtools::check()`.

Reply:
let me clarify this - it might not be super clear at the very first glance and there are a few things to consider
- `shinyApp()` returns an object of class `shiny.appobj`
- `shiny` registers a dedicated `print()` method for this class:
  ```
  r$> shiny:::print.shiny.appobj
  function (x, ...) 
  {
      runApp(x)
  }
  <bytecode: 0x16154bf98>
  <environment: namespace:shiny>
  ```
  As you can see - it's a very simple method - just `runApp()`.

So let see what happens now
If you do `shinyApp(...)` then you (i) creates an object, (ii) also request to print that object, (iii) print calls `runApp()`
If you do `x <- shinyApp(...)` then you _just_ create an object. No print hence no `runApp()`.

In our docs we more or less do the following: `app <- teal::init(); if(interactive()) shinyApp(app)`. That means that if interactive - we call print of the `shiny.appobj` object which then calls `runApp()` as a result. That's to show you that we actually do call `runApp()` - but not explicitly. This is fine as that's the common practice and quite simple workflow - note that we are in the context of examples and examples must be simple enough.

---

This is important in conjunction with `run_examples(..., quiet = TRUE)`. It seems to me that when `quiet = TRUE` then we just _eval_ the code and `print()` is not called. Therefore `runApp()` is not called. I realized about this mistake of mine quite recently and changed it to explicit `quiet = FALSE`. Added a small comment as well.
I am also thinking whether `quiet = TRUE` (i.e. no print) would work if we mock `shiny::shinyApp` instead of `shiny::runApp`. This seems to be independent on the `print()` anymore. On the other hand - this might be bad because `shinytest2` seems to be a wrapper on the top of `runApp()` - it accepts `shiny.appobj` object and creates a separate handler for it. It's arguments relates more to `runApp()` and not `shinyApp()`. 

-- 

You are not seeing this because of `capture.output()` somewhere above. This is needed because `run_examples(..., quiet = TRUE)` would print the whole example source(!) code (incl. comments) to the console and I want to stick to the testthat2 output only. Feel free to disable it temporarily to see this.
To make it even more complex: shinytest2 has its own log and there you can see console output got from evaluation of example source code. If you will put a debug print in the roxygen example code - you would have to call `app_driver$get_logs()`.
(I have removed this debug statement)

--

To sum up. This is a combination of (i) `capture.output()` (ii) `run_examples(..., quiet = TRUE)` (iii) `runApp()` (and  the fact that its called when printing) and (iv) the way how we write our examples. Each of these has some reasons to me. I hope this is a little bit cleaner.
I was targeting a simple and light tool but unfortunately the matter is not super simple and requires a very well thought through specification with some interdepencies.

Reply:
Thanks for the explanation. I'm clear on this.